pkg/kubelet/status/status_manager.go

```diff
@@ -1023,23 +1023,26 @@ func normalizeStatus(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
 		normalizeTimeStamp(&condition.LastTransitionTime)
 	}
 
-	// update container statuses
-	for i := range status.ContainerStatuses {
-		cstatus := &status.ContainerStatuses[i]
-		normalizeContainerState(&cstatus.State)
-		normalizeContainerState(&cstatus.LastTerminationState)
+	normalizeContainerStatuses := func(containerStatuses []v1.ContainerStatus) {
+		for i := range containerStatuses {
+			cstatus := &containerStatuses[i]
+			normalizeContainerState(&cstatus.State)
+			normalizeContainerState(&cstatus.LastTerminationState)
+		}
+		// Sort the container statuses, so that the order won't affect the result of comparison
+		sort.Sort(kubetypes.SortedContainerStatuses(containerStatuses))
 	}
-	// Sort the container statuses, so that the order won't affect the result of comparison
-	sort.Sort(kubetypes.SortedContainerStatuses(status.ContainerStatuses))
 
-	// update init container statuses
-	for i := range status.InitContainerStatuses {
-		cstatus := &status.InitContainerStatuses[i]
-		normalizeContainerState(&cstatus.State)
-		normalizeContainerState(&cstatus.LastTerminationState)
-	}
-	// Sort the container statuses, so that the order won't affect the result of comparison
+	// Normalize and sort container statuses
+	normalizeContainerStatuses(status.ContainerStatuses)
+
+	// Normalize and sort init container statuses
+	normalizeContainerStatuses(status.InitContainerStatuses)
 	kubetypes.SortInitContainerStatuses(pod, status.InitContainerStatuses)
```

Comment:
I thought you moved sorting to `normalizeContainerStatuses`? Am I missing something?

Reply:
> I thought you moved sorting to `normalizeContainerStatuses`? Am I missing something?

Yes, this does lead to an unnecessary sorting of InitContainerStatuses. Would it be better to take the sorting out of normalizeContainerStatuses?

Reply:
either remove it from normalizeContainerStatuses and add to ephemeral or delete from here. I think deleting it from here is the best

Reply:
> either remove it from normalizeContainerStatuses and add to ephemeral or delete from here. I think deleting it from here is the best

Do you prefer deleting `kubetypes.SortInitContainerStatuses`? Would this affect the order in which initContainers's statuses are displayed?

Reply:
You have introduced a new func `normalizeContainerStatuses` that normilizes time and then sorts the collection. You removed the loop that normilizes the time on init contianers. But kept sorting here. So now init containers will be sorted once in `normalizeContainerStatuses`  and then again here. I suggest to remove it from here

Reply:
Also make sure there is a test for forting of statuses

Reply:
`kubetypes.SortInitContainerStatuses` and `sort.Sort` have different sorting rules. `kubetypes.SortInitContainerStatuses` sorts initContainerStatuses based on the order initContainers were created, while `sort.Sort` sorts containers by their names. I'm concerned that if `kubetypes.SortInitContainerStatuses` is removed, it might affect the order of initContainerStatuses. `normalizeStatus` is called within `updateStatusInternal`, and the normalized status is then added to `podStatuses`. The `statusManager` will synchronize `podStatuses` to the apiserver.

https://github.com/kubernetes/kubernetes/blob/f0f7ff989a948389247e628c4c5a43e915f51daa/pkg/kubelet/status/status_manager.go#L654

Reply:
I see. Let's remove the sort from the `normalizeContainerStatuses` and call sort for each collection separately. I wonder how they needs to be sorted for the ephemeral containers

Reply:
> I see. Let's remove the sort from the `normalizeContainerStatuses` and call sort for each collection separately. I wonder how they needs to be sorted for the ephemeral containers

The sorting rule for ephemeral containers is the same as that for regular containers, which is fine because their startup rules are similar, and neither requires ensuring that one is running before starting the next. However, it's worth noting that the existing ephemeral container statuses in the cluster have not been sorted. Updating this PR will cause the ephemeral container statuses to be sorted, but this isn't a problem.